proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -3821,6 +3832,11 @@ lemma cancelBadgedSends_filterM_helper':
 lemmas cancelBadgedSends_filterM_helper
     = spec [where x=Nil, OF cancelBadgedSends_filterM_helper', simplified]
 
+lemma list_case_return:
+  "(case xs of [] \<Rightarrow> return v | y # ys \<Rightarrow> return (f y ys))
+    = return (case xs of [] \<Rightarrow> v | y # ys \<Rightarrow> f y ys)"
+  by (simp split: list.split)
+
```

Comment:
This lemma should be available in Lib or Monads somewhere already

Reply:
I think this was in Lib before the big change to NonDetMonadLemmaBucket. I can't find it now.

There's also hoare_case_option_wp2 and distinct_tuple_helper which were originally in Lib and I think are now gone, but resurrected in this PR. Maybe I missed the Refine lemmas when I did the PR the other day. It would slow me down even more to add them to Lib on master first, but I could do that if you'd like

Reply:
I'm not sure how easy it would be to rebase the work I've done for a merge over a change to `master`. Sounds bad. But what I could do is merge this PR, make another small PR for Lib on `master`, cherry-pick that Lib PR to `rt`, and then just delete these lemmas in a separate PR

Reply:
`list_case_return` is in `lib/NonDetMonadLemmaBucket` now (because `Lib` doesn't include the monads any more). I think that should be available at that point in the proof, so you should be able to just delete it. If not, then something bad has happened in the merge, because it would mean we lost some lemmas.

`distinct_tuple_helper` is also in `lib/NonDetMonadLemmaBucket` (for the same reason).

`hoare_case_option_wp2` seems to be `rt` only at this point. It's Ok to leave that in just `rt` for now, we can sync up more a bit later.

Reply:
Oh I did mess up that file in the merge. What's the best way of correcting that? Should I redo the merge, with the fixing of the conflicts, and then cherry-pick the last big commit from this branch (deleting the lemmas from Refine that will then be in Lib) and then test?

Reply:
Do you have `git config --global rerere.enabled true` enabled? If yes, then it should remember previous conflict resolutions and redoing the merge is probably the easiest way.

If not, it might be easier to try to do this one file only by looking at both diffs manually. @Xaphiosis or @corlewis might have better tips on that.

Reply:
I think I must have that enabled, because I do remember it resolving previous conflicts automatically. I wonder whether I originally messed this file up and it then applied that resolution, or whether it was the second time around. But I'll try again and see if it's any better

Reply:
Oh, yes, that actually makes complete sense. If it remembers a previous mistake it'll just re-apply that.

Reply:
`git rerere forget FILE-WITH-BAD-MERGE` is probably what you need

Reply:
It miraculously didn't automatically apply a previous resolution for this file, so I think it should be fine this time around. I'll locally run AInvs and Refine and update the PR when they've passed